# 예외 종류에 따른 처리
<br/>

try 블록에는 다양한 종류의 예외가 발생할 수 있다. 이 경우, **다중 catch**를 사용하면 발생하는 예외에 따라 예외 처리 코드를 다르게 작성할 수 있다.
catch 블록의 예외 클래스는 try 블록에서 발생된 예외의 종류를 말하는데, 해당 타입의 예외가 발생하면 catch 블록이 선택되어 실행된다.

<img src="https://github.com/silxbro/java/assets/142463332/1476c3ee-b0e9-466d-bf3f-85dc62072248" width="600" height="260"/><br/>

catch 블록이 여러 개라 할지라도 catch 블록은 **`단 하나만`** 실행된다. 그 이유는 try 블록에서 동시다발적으로 예외가 발생하지 않으며, **하나의 예외가 발생하면 즉시 실행을 멈추고
해당 catch 블록으로 이동**하기 때문이다.

<br/>

다음 예제는 **배열의 인덱스가 초과**되었을 경우 발생하는 **`ArrayIndexOutOfBoundsException`** 과
**숫자타입이 아닐 때 발생**하는 **`NumberFormatException`** 을 각각 다르게 예외 처리한다.
- [ExceptionHandlingExample.java](https://github.com/silxbro/java/blob/main/src/thisisjava/ch11/sec03/exam01/ExceptionHandlingExample.java)

<br/>

처리해야 할 예외 클래스들이 상속 관계에 있을 때는 **하위 클래스 catch 블록을 먼저 작성**하고 상위 클래스 catch 블록을 나중에 작성해야 한다.
예외가 발생하면 catch 블록은 위에서부터 차례대로 검사 대상이 되는데, 하위 예외도 상위 클래스 타입이므로 상위 클래스 catch 블록이 먼저 검사 대상이 되면 안 된다.
- [ExceptionHandlingExample.java](https://github.com/silxbro/java/blob/main/src/thisisjava/ch11/sec03/exam02/ExceptionHandlingExample.java)

<br/>

**두 개 이상의 예외를 하나의 catch 블록으로 동일하게 예외 처리**하고 싶을 때가 있다. 이 경우에는 catch 블록에 **예외 클래스를 기호 `|`로 연결**하면 된다.
- [ExceptionHandlingExample.java](https://github.com/silxbro/java/blob/main/src/thisisjava/ch11/sec03/exam03/ExceptionHandlingExample.java)
